### PR TITLE
Fix premature Close() of docker client

### PR DESCRIPTION
**Pull request by matejvasek**
# Changes
* Fix premature Close() of docker client

/kind bug

```release-note
fix: s2i+podman "Cannot connect to the Docker daemon" error
```

### CODE REVIEW

File: s2i/builder.go [hunk 1]
```diff
@@ -115,7 +115,19 @@ func (b *Builder) Build(ctx context.Context, f fn.Function) (err error) {
 
 	// Create the S2I builder instance if not overridden
 	if b.impl == nil {
-		if b.impl, err = newImpl(ctx, cfg); err != nil {
+		var client dockerClient.CommonAPIClient
+		client, _, err = docker.NewClient(dockerClient.DefaultDockerHost)
+		if err != nil {
+			return
+		}
+		defer client.Close()
```
Comment (matejvasek) [inline s2i/builder.go:123]: This is first time I find `defer` scoping of `Go` useful.

**Comment by matejvasek** [comment]
@lance we might wan't to cherry pick this to  0.24.0

**Review by lance — COMMENTED** [review]
/lgtm
/hold for tests

**Pull request merged**